cmd/sdk-server/main.go

```diff
@@ -0,0 +1,234 @@
+
+	c *cache.MemCache
+	g apiserv.Group
+}
+
+func (ch *clientHandler) getClient(ctx *apiserv.Context) (c *sdk.Client) {
+	defer func() {
+		if x := recover(); x != nil {
+			ctx.JSON(http.StatusInternalServerError, true, x)
+		}
+	}()
+
+	key := ctx.Query("apiKey")
+	if key == "" {
+		key = ctx.Header().Get("X-APIKEY")
+	}
+
+	if key == "" {
+		ctx.JSON(http.StatusUnauthorized, true, apiserv.NewJSONErrorResponse(http.StatusUnauthorized, "missing or invalid api key"))
+		return
+	}
+
+	if c, ok := ch.c.Get(key); ok {
+		return c.(*sdk.Client)
+	}
+
+	var err error
+	ch.c.Update(key, func(old interface{}) (_ interface{}, _ bool, _ time.Duration) {
+		if c, _ = old.(*sdk.Client); c != nil {
+			return c, false, time.Hour * 24
+		}
+
+		c = sdk.NewWithAddr(ch.addr, key)
+		var uid string
+		if uid, err = c.GetUserID(context.Background()); err != nil {
+			return
+		}
+
+		// ghetto check, this needs to be handled by an actual endpoint
+		if _, err = c.ListAdvertisers(context.Background(), uid); err != nil {
+			return
+		}
+
+		return c, false, clientCacheTimeout
+	})
+
+	if err != nil {
+		ctx.JSON(http.StatusUnauthorized, true, apiserv.NewJSONErrorResponse(http.StatusUnauthorized, err))
+		c = nil
+	}
+	return
+}
+
+func (ch *clientHandler) GetCampaignReport(ctx *apiserv.Context) apiserv.Response {
+	c := ch.getClient(ctx)
+	if ctx.Done() {
+		return nil
+	}
+	var (
+		callsKey             = fmt.Sprintf("client:%p", c)
+		uid, cid, start, end = ctx.Param("uid"), ctx.Param("cid"), ctx.Param("start"), ctx.Param("end")
+		calls                uint64
+	)
+
+	ch.c.Update(callsKey, func(old interface{}) (_ interface{}, _ bool, _ time.Duration) {
+		calls, _ := old.(uint64)
+		calls++
+		return calls, true, time.Hour
+	})
+
+	if calls > maxReportingCallsPerHour {
+		return apiserv.NewJSONErrorResponse(http.StatusTooManyRequests,
+			fmt.Sprintf("you went over your hourly request limit of %d by %d",
+				maxReportingCallsPerHour, calls-maxReportingCallsPerHour,
+			))
+	}
+
+	cacheKey := cache.Key(uid, cid, start, end)
+	if data, ok := ch.c.Get(cacheKey); ok {
+		return apiserv.NewJSONResponse(data)
+	}
+
+	var (
+		data interface{}
+		err  error
+	)
+
+	ch.c.Update(cacheKey, func(old interface{}) (_ interface{}, _ bool, _ time.Duration) {
+		if data = old; data == nil {
+			data, err = c.GetCampaignReport(context.Background(), uid, cid, sdk.DateToTime(start), sdk.DateToTime(end))
+		}
+
+		return data, false, time.Hour * 3
+	})
+
+	if err != nil {
+		return apiserv.NewJSONErrorResponse(http.StatusBadRequest, err)
+	}
+
+	return apiserv.NewJSONResponse(data)
+}
+
+func (ch *clientHandler) GetAdsReport(ctx *apiserv.Context) apiserv.Response {
+	c := ch.getClient(ctx)
+	if ctx.Done() {
+		return nil
+	}
+
+	var (
+		callsKey        = fmt.Sprintf("client:%p", c)
+		uid, start, end = ctx.Param("uid"), ctx.Param("start"), ctx.Param("end")
+		calls           uint64
+	)
+
+	ch.c.Update(callsKey, func(old interface{}) (_ interface{}, _ bool, _ time.Duration) {
```

Comment:
We have this Update.. then calls> maxReportingCallsPerHour logic copy pasted. Should have a helper func for this we call..would be useful for future reporting hiandlers